client/src/components/modal-forms/BulkAllocationForm.js

```diff
@@ -0,0 +1,164 @@
+import React, { useState } from 'react';
+import { Button, Dialog } from '../generic';
+import { Box, Typography } from '@material-ui/core';
+import Alert from '@material-ui/lab/Alert';
+import { makeStyles } from '@material-ui/core/styles';
+import { BulkAllocationSchema, ToastStatus } from '../../constants';
+import { RenderTextField, RenderSelectField, RenderCheckbox } from '../fields';
+import { Field, Formik, Form as FormikForm } from 'formik';
+import { dateToTextString } from '../../utils/date';
+import { bulkAllocation } from '../../services/allocations';
+import { useToast } from '../../hooks';
+
+const useStyles = makeStyles(() => ({
+  formButton: {
+    maxWidth: '200px',
+  },
+  formRow: {
+    gap: '25px',
+  },
+  list: {
+    maxHeight: '100px',
+    overflow: 'auto',
+    overflowX: 'hidden',
+  },
+}));
+
+export const createPhaseDropdown = (phases) => {
+  return phases
+    .sort((a, b) => new Date(b.end_date) - new Date(a.end_date))
+    .map((phase) => ({
+      value: phase.id,
+      label: `${phase.name} - ${dateToTextString(phase.start_date)} -
+    ${dateToTextString(phase.end_date)}`,
+    }));
+};
+
+export const BulkAllocationForm = ({ onClose, afterSubmit, open, sites, phases = [] }) => {
+  const { openToast } = useToast();
+  const classes = useStyles();
+  const [existingAllocations, setExistingAllocations] = useState([]);
+  const siteIds = sites.map(({ id }) => id);
+  const initialValues = {
+    allocation: '',
+    phase_id: '',
+    existingAllocations: false,
+    acknowledgement: false,
+  };
+
+  const handleSetAllocation = (phaseId, setValue) => {
+    // get all allocations associated to the selected phase
+    // filter the allocations and return all that are associated to the sites selected.
+    const allocations = phases
```

Comment:
This looks like it could be potentially optimized further. Lots of array functions going on that makes it a little harder to read and I wonder if it could be shortened. Don't know how yet, but I'm wondering if maybe some sort of dictionary of allocations that we could pull from.

Feel free to disagree though!

Reply:
It can be shortened. It does require two array.filters/finds. I have split it into two consts to make it easier to read: 
```
 // get all allocations associated to the selected phase
    const phaseAllocations = phases.find(({ id }) => id === phaseId)?.allocations;
    // filter the allocations and return all that are associated to the sites selected.
    const phaseSiteAllocations = phaseAllocations.filter(({ site_id }) =>
      siteIds.includes(site_id)
    );
```

I was then mapping to format the object as I needed the site name to be included in the object, (Initially, I was passing this array to the yup validator/BE so I intentionally made the object smaller)  But that isn't necessary anymore, and I can read the siteName a different way. (created an object hash)